File: artifacts/api-server/src/services/audio-store.ts
/**
 * audio-store.ts — In-memory detection store for the Berkeley audio network.
 *
 * Receives data via POST /api/ingest/audio-<nodeId> (the existing generic
 * ingest endpoint). Data is indexed here for structured REST access via
 * /api/audio/* routes and re-broadcast over the Socket.IO WebSocket hub.
 *
 * This store is the TypeScript counterpart to the Python-side audio_store.py.
 * The Python store is ephemeral per-process; this store is the authoritative
 * source for all HTTP/WebSocket consumers (Berkeley display, future audio
 * monitor, etc.).
 *
 * WebSocket channels emitted:
 *   audio:detection   — single detection object, broadcast on every ingest
 *   audio:node:status — node status change (online/offline/degraded)
 *
 * All data is in-memory only — no database writes. The api-server is
 * stateless by design. Persistent storage lives in the archived WAV files
 * written by audio_archiver.py.
 */

import { broadcast } from "./ws-hub";
import { logger } from "../lib/logger";

// ── Types ─────────────────────────────────────────────────────────────────────

export type Analyzer = "birdnet" | "batnet" | string; // extensible

export interface NodeLocation {
  name: string;
  description: string;
  lat: number;
  lng: number;
}

export interface AudioDetection {
  nodeId: string;
  location: NodeLocation;
  timestamp: number;      // epoch ms, set by Python poster
  analyzer: Analyzer;
  species: string;        // scientific name
  commonName: string;
  confidence: number;     // 0.0–1.0
  startTime: number;      // seconds into the audio chunk
  endTime: number;
  chunkFile?: string;     // archive path if clip was saved
}

export interface NodeStatus {
  nodeId: string;
  status: "online" | "offline" | "degraded" | "initializing";
  detail: string;
  lastSeen: number;       // epoch ms
}

// ── Constants ─────────────────────────────────────────────────────────────────

/** Max detections kept per node (ring buffer) */
const MAX_PER_NODE = 500;

/** Max entries in the global cross-node feed */
const MAX_GLOBAL = 2000;

// ── Internal state ─────────────────────────────────────────────────────────────

const detectionsByNode = new Map<string, AudioDetection[]>();
const globalFeed: AudioDetection[] = [];
const nodeStatuses = new Map<string, NodeStatus>();

// ── Ingest ────────────────────────────────────────────────────────────────────

/**
 * Called by the /api/ingest route handler for any `audio-*` service name.
 * Parses the raw ingest payload, indexes detections, and broadcasts to WS.
 */
export function ingestAudioPayload(nodeId: string, data: any, metadata?: any): void {
  // ── Node status update ────────────────────────────────────────────────────
  if (metadata?.type === "status") {
    updateNodeStatus(nodeId, data.status ?? "online", data.detail ?? "");
    return;
  }

  // ── Detection batch ───────────────────────────────────────────────────────
  const analyzer: Analyzer = data.analyzer ?? "birdnet";
  const location: NodeLocation = data.location ?? { name: nodeId, description: "", lat: 0, lng: 0 };
  const rawDetections: any[] = Array.isArray(data.detections) ? data.detections : [];

  if (rawDetections.length === 0) return;

  const ts = typeof data.timestamp === "number" ? data.timestamp : Date.now();

  const detections: AudioDetection[] = rawDetections.map((d) => ({
    nodeId,
    location,
    timestamp: ts,
    analyzer,
    species: String(d.species ?? ""),
    commonName: String(d.commonName ?? ""),
    confidence: Number(d.confidence ?? 0),
    startTime: Number(d.startTime ?? 0),
    endTime: Number(d.endTime ?? 0),
    chunkFile: d.chunkFile,
  }));

  // Store per-node
  if (!detectionsByNode.has(nodeId)) {
    detectionsByNode.set(nodeId, []);
  }
  const nodeBuf = detectionsByNode.get(nodeId)!;
  nodeBuf.push(...detections);
  while (nodeBuf.length > MAX_PER_NODE) nodeBuf.shift();

  // Store in global feed
  globalFeed.push(...detections);
  while (globalFeed.length > MAX_GLOBAL) globalFeed.shift();

  // Mark node as online on any successful ingest
  updateNodeStatus(nodeId, "online");

  // Broadcast each detection individually for live feeds
  for (const det of detections) {
    broadcast("audio:detection", det);
  }

  logger.debug({ nodeId, analyzer, count: detections.length }, "Audio detections indexed");
}

// ── Node status ───────────────────────────────────────────────────────────────

export function updateNodeStatus(
  nodeId: string,
  status: NodeStatus["status"],
  detail = "",
): void {
  const existing = nodeStatuses.get(nodeId);
  const updated: NodeStatus = {
    nodeId,
    status,
    detail,
    lastSeen: Date.now(),
  };
  nodeStatuses.set(nodeId, updated);

  if (existing?.status !== status) {
    broadcast("audio:node:status", updated);
    logger.info({ nodeId, status, detail }, "Audio node status changed");
  }
}

// ── Queries ───────────────────────────────────────────────────────────────────

export function getNodeDetections(nodeId: string, limit = 50): AudioDetection[] {
  const buf = detectionsByNode.get(nodeId) ?? [];
  return buf.slice(-Math.min(limit, MAX_PER_NODE));
}

export function getGlobalFeed(limit = 100): AudioDetection[] {
  return globalFeed.slice(-Math.min(limit, MAX_GLOBAL));
}

export function getAllNodeStatuses(): NodeStatus[] {
  return Array.from(nodeStatuses.values());
}

export function getNodeStatus(nodeId: string): NodeStatus | undefined {
  return nodeStatuses.get(nodeId);
}

export function getActiveNodeIds(): string[] {
  return Array.from(detectionsByNode.keys());
}

/**
 * Summary stats per node — used by /api/audio/nodes for a status overview.
 */
export function getNodeSummaries() {
  return Array.from(detectionsByNode.entries()).map(([nodeId, dets]) => {
    const status = nodeStatuses.get(nodeId);
    const recent = dets.slice(-10);
    const topSpecies = _topSpecies(recent);
    return {
      nodeId,
      status: status?.status ?? "unknown",
      lastSeen: status?.lastSeen ?? 0,
      totalDetections: dets.length,
      recentDetections: recent.length,
      topSpecies,
    };
  });
}

function _topSpecies(dets: AudioDetection[]): { commonName: string; count: number }[] {
  const counts = new Map<string, number>();
  for (const d of dets) {
    counts.set(d.commonName, (counts.get(d.commonName) ?? 0) + 1);
  }
  return Array.from(counts.entries())
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5)
    .map(([commonName, count]) => ({ commonName, count }));
}

File: audio-receiver/src/audio_store.py
"""
audio_store.py — Local in-process detection ring-buffer.

Provides a thread-safe store for recent detections per node.
This is the Python-side cache; the api-server maintains its own
TypeScript-side store for API/WebSocket consumers.
"""

import threading
import time
from collections import deque
from typing import Any

from logger import get_logger

log = get_logger("audio_store")

_MAX_PER_NODE = 500   # detections to keep per node
_MAX_GLOBAL = 2000    # global rolling window


class AudioStore:
    """Thread-safe in-memory store for recent audio detections."""

    def __init__(self, max_per_node: int = _MAX_PER_NODE, max_global: int = _MAX_GLOBAL):
        self._lock = threading.Lock()
        self._by_node: dict[str, deque] = {}
        self._global: deque = deque(maxlen=max_global)
        self._max_per_node = max_per_node

    def add(
        self,
        node_id: str,
        analyzer: str,
        detections: list[dict[str, Any]],
        node_meta: dict[str, Any],
        chunk_file: str | None = None,
    ) -> None:
        """Store a batch of detections from one analysis run."""
        ts = int(time.time() * 1000)

        records = []
        for det in detections:
            record = {
                "nodeId": node_id,
                "location": node_meta.get("location_obj", {}),
                "timestamp": ts,
                "analyzer": analyzer,
                "species": det.get("species", ""),
                "commonName": det.get("commonName", ""),
                "confidence": det.get("confidence", 0.0),
                "startTime": det.get("startTime", 0.0),
                "endTime": det.get("endTime", 0.0),
                "chunkFile": chunk_file,
            }
            records.append(record)

        with self._lock:
            if node_id not in self._by_node:
                self._by_node[node_id] = deque(maxlen=self._max_per_node)
            for rec in records:
                self._by_node[node_id].append(rec)
                self._global.append(rec)

        if records:
            log.debug("Stored detections", extra={
                "node": node_id,
                "analyzer": analyzer,
                "count": len(records),
            })

    def get_node(self, node_id: str, limit: int = 50) -> list[dict]:
        """Return recent detections for a specific node."""
        with self._lock:
            buf = self._by_node.get(node_id, deque())
            items = list(buf)
        return items[-limit:]

    def get_global(self, limit: int = 100) -> list[dict]:
        """Return most recent detections across all nodes."""
        with self._lock:
            items = list(self._global)
        return items[-limit:]

    def get_node_ids(self) -> list[str]:
        """Return list of node IDs that have received detections."""
        with self._lock:
            return list(self._by_node.keys())

    def clear_node(self, node_id: str) -> None:
        with self._lock:
            if node_id in self._by_node:
                self._by_node[node_id].clear()


# Singleton instance used across the service
store = AudioStore()
